docs: migrate classic linting how-to (#5415)
canonical/snapcraft: docs/how-to/debug/use-the-classic-linter.rst

Use the classic linter
======================

Snapcraft uses :ref:`linters <reference-linters>` to check for issues during builds.

> Linters can only be specified in snaps that use the core22 base or higher. Warnings

Reviewer: ```suggestion
Linters can only be specified in snaps that use core22 or higher. Warnings
```

> are still reported for snaps that use older bases.

Reviewer: I feel like this paragraph belongs in the linter index.

Author: Looks like this language already more or less existed in reference/linters.rst, so I'm just going to remove it here. Good catch

> The classic linter is a Snapcraft linter that is used to verify binary file parameters

Reviewer: ```suggestion
The *classic linter* in Snapcraft verifies binary file parameters
```

We don't do this everywhere, but this is an effective way to introduce a term.

to ensure they are set appropriately for snaps using
:ref:`explanation-classic-confinement`.

> The classic linter is only invoked when snap confinement is set to ``classic``, or if
> ``libc`` is staged.

Reviewer: ```suggestion
The classic linter is only invoked when a snap has :ref:`classic confinement
<explanation-classic-confinement>` or it stages ``libc``.
```

The classic linter will issue a warning if the ELF binary it is testing either:

> - does not have the correct dynamic linker (either from the base or the staged
>   ``libc``)
> - does not have an ``rpath`` (run-time search path) set to a value needed to load
>   shared library dependencies, either from the base or from the snap if the dependency
>   is not part of the base.

Reviewer: ```suggestion
- Doesn't have the correct dynamic linker (either from the base or the staged
  ``libc``), or
- Doesn't have an ``rpath`` (run-time search path) set to a value needed to load
  shared library dependencies, either from the base or from the snap if the dependency
  is not part of the base.
```

Use the classic linter
======================

Snapcraft uses :ref:`linters <reference-linters>` to check for issues during builds.
Linters can only be specified in snaps that use the core22 base or higher. Warnings
are still reported for snaps that use older bases.

The classic linter is a Snapcraft linter that is used to verify binary file parameters
to ensure they are set appropriately for snaps using
:ref:`explanation-classic-confinement`.

The classic linter is only invoked when snap confinement is set to ``classic``, or if
``libc`` is staged.

The classic linter will issue a warning if the ELF binary it is testing either:

- does not have the correct dynamic linker (either from the base or the staged
  ``libc``)
- does not have an ``rpath`` (run-time search path) set to a value needed to load
  shared library dependencies, either from the base or from the snap if the dependency
  is not part of the base.

See :ref:`how-to-disable-a-linter` for details on how to stop this linter from running.

Find issues at build time
-------------------------

To address classic linter issues, the appropriate ``rpath`` can be set during build

> time, or existing binaries can be patched to have their rpath changed.

Author: I see no problem with changing it to say that, it's probably a lot easier to write snaps doing that anyways.
